Handle Tekton pipeline level variables

jinchihe opened the pull request:
/feature

**Which issue is resolved by this Pull Request:** 

The PR is going to resolve the problem:

If we used the tekton pipeline level variables,  such as `context.pipelineRun.name` or `context.pipelineRun.namespace`, see detail of tekton variable [here](https://github.com/tektoncd/pipeline/blob/master/docs/variables.md) . 
Currently even if we defined those variables in inputs section of components,  after compiled by dsl, the variables is not loaded, and only replace value in containerOps. In fact, that's pipeline level variables, we need to pass that by `params`, cannot use ``context.pipelineRun.namespace` in containerOps directly, see the example for details: https://github.com/tektoncd/pipeline/blob/master/examples/v1beta1/pipelineruns/using_context_variables.yaml

**Description of your changes:**

If there is tekton pipeline level variables in container, then
1. Replease the `$(context.pipeline/pipelineRun).*)` to `$(params.var_name)`, for example: `context.pipelineRun.name` -> `$(params.pipelineRun-name)`
2. Add `$(context.pipeline/pipelineRun).*)` to `pipeline_run['spec']['pipelineSpec']['tasks']['params']`
3. Add `$(context.pipeline/pipelineRun).*)` to `pipeline_run['spec']['pipelineSpec']['tasks']['taskSpec']['params']`

The the container can use the variables normally.

**Environment tested:**

* Python Version (use `python --version`):
* Tekton Version (use `tkn version`):
* Kubernetes Version (use `kubectl version`):
* OS (e.g. from `/etc/os-release`):

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?
    
    If yes, use one of the following options:
  
    * **(Recommended.)** Ask the PR approver to add the `cherrypick-approved` label to this PR. The release manager adds this PR to the release branch in a batch update.
    *  After this PR is merged, create a cherry-pick PR to add these changes to the release branch. (For more information about creating a cherry-pick PR, see the [Kubeflow Pipelines release guide](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#option--git-cherry-pick).)

jinchihe commented:
/cc @fenglixa @Tomcli 

Would you please review? Great thanks.

jinchihe commented:
@Tomcli Thanks for your reviewing, I would like to add a new test since that's new features, instead of argo variable replacing.

I already added a new tests for the change, and update `sdk/FEATURES.md` to describe that. Thanks.

jinchihe commented:
@Tomcli @fenglixa This is ready for another look.

By the way, shall we release a new minor version (0.5.1?) for recently changes? Thanks.

jinchihe commented:
/cc @vincent-pli

fenglixa commented on sdk/python/tests/compiler/testdata/tekton_pipeline_variables.py:
```diff
@@ -0,0 +1,39 @@
+        arguments=["echo \
+            Pipeline name: $(context.pipeline.name), \
+            PipelineRun name: $(context.pipelineRun.name), \
+            PipelineRun namespace: $(context.pipelineRun.namespace), \
```
Thanks @jinchihe, could you consider the code of here:https://github.com/kubeflow/kfp-tekton/blob/master/sdk/python/kfp_tekton/compiler/compiler.py#L766-L777
and support the case that user already defined the argo variables: `{{workflow.namespace}}`?

jinchihe replied:
Yeah, great comments. I knew that, but the replacing is late while compiling.
Any way, your comments is reasonable, I will handle that. Thanks. @fenglixa

jinchihe commented:
Addressed @fenglixa comments, also updated below tests to verify that:
https://github.com/kubeflow/kfp-tekton/blob/master/sdk/python/tests/compiler/testdata/parallel_join_with_argo_vars.py

fenglixa commented:
/lgtm

@Tomcli , is this OK to your side?

Tomcli commented:
it looks good to me
@ckadner will this be conflicting with your merge to kfp 1.3?

ckadner commented on sdk/python/kfp_tekton/compiler/_tekton_hander.py:
```diff
@@ -0,0 +1,54 @@
+import json
+
+
+def _handle_tekton_pipeline_variabes(pipeline_run):
```
"_variabes" should be "_variables" (missing an L)

jinchihe replied:
Great catch! Done

ckadner commented on sdk/python/kfp_tekton/compiler/_tekton_hander.py:
```diff
@@ -0,0 +1,54 @@
+# Copyright 2019-2020 kubeflow.org.
```
since this file was created in 2021 the copyright should read "Copyright 2021 kubeflow.org."

jinchihe replied:
Done

ckadner commented on sdk/python/kfp_tekton/compiler/_tekton_hander.py:
```diff
@@ -0,0 +1,54 @@
+
+def _handle_tekton_pipeline_variabes(pipeline_run):
+    """
+    Handle tekton pipeline level variables, such as context.pipelineRun.name
```
this file should probably be called `_tekton_variable_handler.py` instead of `_tekton_hander.py `

jinchihe replied:
Thanks @ckadner I would like to keep as `_tekton_hander.py`, discussed with @fenglixa, we want to put all tekton handler in this file in the future, like the variables handler.

ckadner commented on sdk/python/kfp_tekton/compiler/_tekton_hander.py:
```diff
@@ -0,0 +1,54 @@
+    """
+    Handle tekton pipeline level variables, such as context.pipelineRun.name
+    See more here https://github.com/tektoncd/pipeline/blob/master/docs/variables.md
+    If there is tekton pipeline level variables in container, then
```
"is" --> "are"

ckadner commented on sdk/python/tests/compiler/testdata/tekton_pipeline_variables.py:
```diff
@@ -0,0 +1,39 @@
+# Copyright 2020 kubeflow.org
```
2020 --> 2021

ckadner commented on sdk/python/tests/compiler/testdata/tekton_pipeline_variables.yaml:
```diff
@@ -0,0 +1,52 @@
+# Copyright 2020 kubeflow.org
```
2020 --> 2021

jinchihe replied:
Done

ckadner reviewed (changes requested):
Thanks @jinchihe -- I just have a few small wording corrections. Otherwise this looks good.

ckadner commented:
> @ckadner will this be conflicting with your merge to kfp 1.3?

Nope.

jinchihe commented:
Thanks @ckadner for your reviewing. Addressed your comments.

fenglixa commented:
Thanks @jinchihe 
/approve

fenglixa commented:
/lgtm
